macaroon_service.go

```diff
@@ -231,12 +234,46 @@ func (ms *MacaroonService) Start() error {
 
 	// There are situations in which we don't want a macaroon to be created
 	// on disk (for example when running inside LiT stateless integrated
-	// mode). For any other cases, we create macaroon files in the default
-	// directory.
-	if ms.cfg.StatelessInit || lnrpc.FileExists(ms.cfg.MacaroonPath) {
+	// mode).
+	if ms.cfg.StatelessInit {
 		return nil
 	}
 
+	// If there are no required permissions, then we don't have to generate
+	// or re-generate a macaroon and so we can exit here.
+	if len(ms.cfg.RequiredPerms) == 0 {
+		return nil
+	}
+
+	// If we are not in stateless mode and a macaroon file does exist, we
+	// check that the macaroon matches the required permissions. If not, we
+	// will delete the macaroon and create a new one.
+	if lnrpc.FileExists(ms.cfg.MacaroonPath) {
+		matches, err := ms.macaroonMatchesPermissions()
+		if err != nil {
+			return fmt.Errorf("unable to match the previous "+
+				"macaroon's permissions with the required "+
+				"permissions: %v", err)
+		}
+
+		// In case the old macaroon matches the required permissions,
+		// we don't need to create a new macaroon.
+		if matches {
+			return nil
+		}
+
+		// Else if the required permissions have been updated, we delete
+		// the old macaroon and create a new one.
+		log.Infof("Macaroon at %s does not have all required "+
+			"permissions. Deleting it and creating a new "+
+			"one", ms.cfg.MacaroonPath)
+
+		if err := os.Remove(ms.cfg.MacaroonPath); err != nil {
```

Comment:
I think perhaps better not to remove the file and just allow `os.WriteFile` below to overwrite the file (it does that by default). Otherwise there is an intermediary state here where perhaps other services making use of this macaroon for calls all of a sudden cant find it.

Reply:
no i mean another service reading the macaroon to make calls. Not writing to it

Reply:
Oh yeah makes sense, thanks for the clarification!